src/libraries/Common/tests/System/Net/Prerequisites/RemoteLoopServer/Handlers/RemoteLoopHandler.cs

```diff
@@ -0,0 +1,155 @@
+
+        public static async Task InvokeAsync(HttpContext context)
+        {
+            try
+            {
+                if (!context.WebSockets.IsWebSocketRequest)
+                {
+                    context.Response.StatusCode = 400;
+                    context.Response.ContentType = "text/plain";
+                    await context.Response.WriteAsync("Not a websocket request");
+
+                    return;
+                }
+
+                using (WebSocket socket = await context.WebSockets.AcceptWebSocketAsync())
+                {
+                    await ProcessWebSocketRequest(context, socket);
+                }
+
+            }
+            catch (Exception)
+            {
+                // We might want to log these exceptions. But for now we ignore them.
+            }
+        }
+
+        private static async Task ProcessWebSocketRequest(HttpContext context, WebSocket control)
+        {
+            byte[] controlBuffer = new byte[128 * 1024];
+            byte[] testedBuffer = new byte[128 * 1024];
+            Socket listenSocket = null;
+            Socket tested = null;
+            CancellationTokenSource cts = new CancellationTokenSource();
+            try
+            {
+                WebSocketReceiveResult first = await control.ReceiveAsync(controlBuffer, cts.Token).ConfigureAwait(false);
+
+                // parse request
+                var message = Encoding.ASCII.GetString(controlBuffer, 0, first.Count);
+                var split = message.Split(',');
+                var listenBacklog = int.Parse(split[0]);
+                var address = IPAddress.Parse(split[1]);
+
+                listenSocket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                listenSocket.Bind(new IPEndPoint(address, 0));
+                listenSocket.Listen(listenBacklog);
+                EndPoint endPoint = listenSocket.RemoteEndPoint ?? listenSocket.LocalEndPoint;
+
+                // respond with what we have done
+                await control.SendAsync(Encoding.ASCII.GetBytes(endPoint.ToString()), WebSocketMessageType.Binary, true, cts.Token).ConfigureAwait(false);
+
+                // wait for the tested client to connect
+                tested = await listenSocket.AcceptAsync().ConfigureAwait(false);
+
+                Task<int> testedNext = tested.ReceiveAsync(new Memory<byte>(testedBuffer), SocketFlags.None, cts.Token).AsTask();
+                Task<WebSocketReceiveResult> controlNext = control.ReceiveAsync(controlBuffer, cts.Token);
+
+                // now we are connected, pump messages
+                bool close = false;
+                while (!close)
+                {
+                    // wait for either message
+                    await Task.WhenAny(testedNext, controlNext).ConfigureAwait(false);
+
+                    if (testedNext.IsCompleted)
+                    {
+                        if (testedNext.Result > 0)
```

Comment:
Might be better to check the task state, and connected status before using the Result.

Reply:
It could be in Close and with data at the same time. This order works better.

Reply:
fixed, thanks